### PR TITLE
Tests for transaction validation

### PR DESCRIPTION
**Pull request by nickchapman-da**
Testing for transaction validation (`isReplayedBy`)

- `Tweak` based testing
- methodically tweak every field of every kind of `GenActionNode`
- classifying each tweak as SIGNIFICANT or INSIGNIFICANT
- test over a range of hand constructed sample transactions (no scalacheck here!)

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+      ne.exerciseResult match {
+        case None => Nil
+        case Some(v) =>
+          List(
+            ne.copy(exerciseResult = Some(changeValue(v))),
+            ne.copy(exerciseResult = None),
+          )
+      }
+    case _ => Nil
+  }
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:528]: With the `PartialFunction` change for `Tweak.apply`

```suggestion
  private val tweakExerciseExerciseResultSome = Tweak[Node] { //sig
    case ne: Node.NodeExercises[_, _] if ne.exerciseResult.isDefined =>
          List(
            ne.copy(exerciseResult = ne.exerciseResult.map(changeValue)),
            ne.copy(exerciseResult = None),
          )
      }
  }
```

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+    case ne: Node.NodeExercises[_, _] =>
+      tweakOptKeyMaintainers.run(ne.key).map { x => ne.copy(key = x) }
+    case _ => Nil
+  }
+  private def tweakExerciseByKey(whenVersion: TransactionVersion => Boolean) = Tweak.single[Node] {
+    case ne: Node.NodeExercises[_, _] =>
+      if (whenVersion(ne.version)) Some(ne.copy(byKey = changeBoolean(ne.byKey)))
+      else None
+    case _ => None
+  }
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:547]: with the `PartialFunction` change in `Tweak.single`.

```suggestion
  private def tweakExerciseByKey(whenVersion: TransactionVersion => Boolean) = Tweak.single[Node] {
    case ne: Node.NodeExercises[_, _] if whenVersion(ne.version) =>
      ne.copy(byKey = changeBoolean(ne.byKey))
  }
```

**Review by remyhaemmerle-da — APPROVED** [review]
I like it very much. Thanks a lot.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+      def f(x: X) = f1(x) match {
+        case None => Nil
+        case Some(y) => List(y)
+      }
+      new Tweak(f)
+    }
+    def apply[X](f: X => List[X]): Tweak[X] = {
+      new Tweak(f)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:64]: partial function to list seems a bit odd. Afaict we have no difference between returning an empty list and the function not being defined.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+    x + "_XXX"
+  }
+
+  //--[predicates]--
+  // Some tweaks have version dependant significance.
+
+  val versionBeforeMinByKey: (TransactionVersion => Boolean) = { v =>
+    import scala.Ordering.Implicits.infixOrderingOps
+    v < TransactionVersion.minByKey
+  }
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: val? -- no good reason. I changed to def.
not private? -- oversight. fixed now.
(good spot, thanks)
the reason I liked to use `private` here (when I remembered!) is that is ensures that everything definition has a use.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:64]: `new Tweak(f.orElse(_ => List.empty[X]))` may work better.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+  //--[shared sub tweaks]--
+
+  private val tweakPartySet = Tweak[Set[Party]] { xs =>
+    (xs + samPartyX) ::
+      List(samParties1, samParties2, samParties3, samParties4).filter(set => set != xs)
+  }
+
+  private def tweakKeyMaintainers = Tweak[KWM] { x =>
+    List(samKWM1, samKWM2, samKWM3).filter(y => x != y)
+  }
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: not a val? -- no good reason. changed to a `val`.

If you want to know the actual reason... during development I often use `???` for code I haven''t yet written, but I often want to check the overall structure compiles. However, `scalac` complains about unreachable code for combinations of `val`/`???`, so I must use `def` as a temporary workaround.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+      count += n
+      assert(n > 0) // ensure tweak actualy applies to something
+      s"[#$n] $name" in {
+        pairs.foreach { case (txA, txB) =>
+          val res = Validation.isReplayedBy(txA, txB)
+          res match {
+            case Right(()) => ()
+            case Left(_) =>
+              println(s"**expected $name to be INSIGNIFICANT (but it wasn't)\n-txA=$txA\n-txB=$txB")
+              assert(false)
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: ok, I used fail

**Review by cocreature — APPROVED** [review]
Nice, I think there is a set of tests that is missing here that we should add but doesn’t have to be in this PR and some of this might also become redundant by adding the normalization:

Anything around items having different numbers of children: e.g., different number of children in an exercise rollback but also different number of record fields. More generally, I think there are probably a few more tests around values (e.g., tweak of record constructor) that could be useful.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+          }
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: because it is quite helpful to be able to see the `txA` and `txB`

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:64]: Thank you for the suggestion about `PartialFunction`. I am aware of it, but to be honest it's not something that is in my everyday scala toolbox. Also, I would be a bit unhappy to loose exhaustiveness checking for pattern matches. However, perhaps it would be idiomatic for the use case here. I wont make the change in this PR. Let's discuss offline & you can try to persuade me. :)

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: :+1:

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+  private def insignificantTweaks: Map[String, Tweak[VTX]] = {
+    (insigCreateTweaks ++ insigFetchTweaks ++ insigLookupTweaks ++ insigExeTweaks)
+      .map { case (name, tw) => (name, tweakTxNodes(tw)) }
+  }
+
+  //--[per tweak tests]--
+
+  "Significant tweaks" - {
+    var count = 0
+    significantTweaks.foreach { case (name, tweak) =>
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:609]: I'll have a look at this...

**Comment by nickchapman-da** [comment]
> partial function to list seems a bit odd. Afaict we have no difference between returning an empty list and the function not being defined.

A Tweak[X] is a total function from `X` to `List[X]`

(1) A non-empty return list provides a list of value-replacements which we want to apply here.
(2) An empty list means, this tweak does not apply here.

There is no concept of a tweak being applicable, but with a empty list of value-replacements. Or at least the meaning of this is indistinguishable from (2).

**Comment by remyhaemmerle-da** [comment]
> > partial function to list seems a bit odd. Afaict we have no difference between returning an empty list and the function not being defined.
> 
> A Tweak[X] is a total function from `X` to `List[X]`

I think Mortiz is referring to **my** suggestion to use `PartialTransaction`.
I do not propose to change Tweak, but the two builders to use `PartialTransaction` instead of total function, specially because You use partial function notation in all your usage of the builders. 

Here is a [concrete example](https://github.com/digital-asset/daml/compare/nick-validation-tests?expand=1) of what I have in mind.


To answer Moritz's concern I do not find to have a "partial function to list" so odd, the idea is just to understand that where the function passed to the builder is undefined, we just use the default value `List.empty`. 
I think this make the writing of test case much lest tedious.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (S11001001) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:64]: I don't see any use for the `Tweak` class. A type synonym will serve equally well.

When you write a function with partial function syntax, _and_ the _expected type_ is a _non_-partial function, you will still get exhaustiveness checking. Broadly speaking, this violates the Liskov Substitution Principle (because an expression that satisfies type `t` does not satisfy type `u >: t` if warnings are errors), but not in a way that violates soundness.

So that means you can write

```scala

private type Tweak[T] = T => List[T]
private object Tweak {
  def apply[T](f: Tweak[T]): Tweak[T] = f
}

Tweak[Option[Location]] {
  case None => //blah blah
  // I get an exhaustiveness warning if I don't handle Some
}
```

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 9]
```diff
@@ -0,0 +1,649 @@
...
+  private val samValue2: Val = V.ValueContractId(samContractId1)
+
+  private val samKWM1 = KeyWithMaintainers(samValue1, samParties1)
+  private val samKWM2 = KeyWithMaintainers(samValue1, samParties2)
+  private val samKWM3 = KeyWithMaintainers(samValue2, samParties1)
+
+  private val samVersion1: TransactionVersion = TransactionVersion.minVersion
+  private val samVersion2: TransactionVersion = TransactionVersion.maxVersion
+
+  private val someCreates: List[Node] =
```
Comment (S11001001) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:120]: This kind of code should broadly use `Seq` instead of `List`. It would probably be profitably idiomatic to replace every `List` with `Seq` here.

The only legitimate reasons to use `List` directly are

1. you need to traverse or
2. need a functor/monad instance, or
3. you rely specifically on the structure of a list, i.e. cheap push/pop

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+  private val tweakOptContractId = Tweak[Option[V.ContractId]] { x =>
+    List(None, Some(samContractId1), Some(samContractId2)).filter(y => x != y)
+  }
+
+  //--[Create node tweaks]--
+
+  private val tweakCreateCoid = Tweak.single[Node] {
+    case nc: Node.NodeCreate[_] => Some(nc.copy(coid = changeContractId(nc.coid)))
+    case _ => None
+  }
```
[the remark this answers is not in the document]
Comment (S11001001) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:299]: It is perfectly reasonable to have another method in `object Tweak` that takes a `PartialFunction` instead and adds the `Seq.empty` case with `orElse`, producing a total function. You are getting no exhaustiveness-checking value out of many of these tweak definitions, and defining them as @remyhaemmerle-da suggests would clean them up.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala [hunk 11]
```diff
@@ -0,0 +1,649 @@
...
+          val res = Validation.isReplayedBy(txA, txB)
+          res match {
+            case Left(_) => ()
+            case Right(()) =>
+              fail(s"**expected $name to be SIGNIFICANT (but it wasn't)\n-txA=$txA\n-txB=$txB")
+          }
+        }
+      }
+    }
+    println(s"**SIGNIFICANT tweaks checked: #$count")
```
Comment (S11001001) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:625]: All `println`s should be removed.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (S11001001) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala:609]: `forEvery` is already doing a lot of the stuff for you so you don't have to do it yourself.  For example, a nested `forEvery` here would simply report the `txA` and `txB` in a nice way as part of the failure message, so you would not have to jam it into the `fail` string.

And _that_ makes it so that you can simply use `shouldBe` or `inside` without using `fail` at all, and still getting all the stuff you might need to know to understand failures reported.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (S11001001) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala]: `forEvery` solves this problem nicely, as mentioned in https://github.com/digital-asset/daml/pull/10167#discussion_r664890874.

But there's another aspect here that really illustrates why it's useful to track down the existing combinators rather than use scala-library to simulate them. There are two kinds of failures of these kinds of tests: break-everything, and break-something. As written, these tests stop on the first failure. So you don't really know which of those situations you're in.

With `forEvery`, you can immediately tell from one test run whether you've broken everything, just a few things, or a single edge case. It's a bit like the difference between having separate test declarations for each test, and just putting every assertion in the whole file into a single test case.

Redefining this functionality is _annoying_. And it is quite useful; #8879 was more or less an exercise of adding cases, tweaking or rewriting part of the algorithm, and quickly getting feedback on every shortcoming of whatever change I had just made--whether there was just one detail I had forgotten, or everything collapsed and I had hit a dead end.

**Comment by nickchapman-da** [comment]
Thanks @stephencompall-DA  for your suggestions. Please see: #10214
